File: LeetCodeProblemsCMake/Problems/LeetCode898BitwiseORsOfSubarrays.cpp
//
// Created by saeed on 7/31/25.
//

#include "LeetCode898BitwiseORsOfSubarrays.h"

#include <unordered_set>
using namespace std;


int LeetCode898BitwiseORsOfSubarrays::subarrayBitwiseORs(vector<int> &arr) {
    unordered_set<int> result{};
    unordered_set<int> current{};
    unordered_set<int> next{};
    for (auto num: arr) {
        next = {};
        next.insert(num);
        for (auto orVal: current) {
            next.insert(orVal | num);
        }
        current = next;
        result.insert(next.begin(), next.end());
    }
    return static_cast<int>( result.size());
}

File: LeetCodeProblemsGo/LeetCode898BitwiseORsOfSubarrays/LeetCode898BitwiseORsOfSubarrays.go
package LeetCode898BitwiseORsOfSubarrays

func subarrayBitwiseORs(arr []int) int {
	result := make(map[int]any)
	current := make(map[int]any)
	for _, num := range arr {
		next := make(map[int]any)
		next[num] = struct{}{}
		for orValue, _ := range current {
			next[orValue|num] = struct{}{}
		}
		current = next
		for orValue, _ := range current {
			result[orValue] = struct{}{}
		}
	}
	return len(result)

}
